caches/infinispan/src/test/java/org/commonjava/maven/galley/cache/routes/RoutingCacheProviderWrapperTest.java

```diff
@@ -0,0 +1,124 @@
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertThat;
+
+public class RoutingCacheProviderWrapperTest
+{
+    @Rule
+    public TemporaryFolder temp = new TemporaryFolder();
+
+    private CacheProviderFactory partylineFac;
+
+    private CacheProviderFactory fastLocalFac;
+
+    private final PathGenerator pathgen = new HashedLocationPathGenerator();
+
+    private final FileEventManager events = new TestFileEventManager();
+
+    private final TransferDecorator decorator = new TestTransferDecorator();
+
+    private final RouteSelector selector = new RouteSelector()
+    {
+        @Override
+        public boolean isDisposable( ConcreteResource resource )
+        {
+            if ( resource != null )
+            {
+                String pattern = "^hosted:.*$";
+                final Location loc = resource.getLocation();
+                if ( loc != null )
+                {
+                    final String uri = loc.getUri();
+                    synchronized ( this )
+                    {
+                        return uri != null && uri.matches( pattern );
+                    }
+                }
+            }
+            return false;
+        }
+    };
+
+    @Before
+    public void prepare()
+            throws Exception
+    {
+
+        final File cacheDir = temp.newFolder();
+        partylineFac = new PartyLineCacheProviderFactory( cacheDir );
+        fastLocalFac = new FastLocalCacheProviderFactory( cacheDir, temp.newFolder(), null,
+                                                          Executors.newFixedThreadPool( 5 ) );
+    }
+
+    @Test
+    public void test()
+            throws Exception
+    {
+
+        final RoutingCacheProviderWrapper router =
+                (RoutingCacheProviderWrapper) new RoutingCacheProviderFactory( selector, fastLocalFac,
+                                                                               partylineFac ).create( pathgen,
+                                                                                                      decorator,
+                                                                                                      events );
+        final CacheProvider partyline = partylineFac.create( pathgen, decorator, events );
+        final CacheProvider fastLocal = fastLocalFac.create( pathgen, decorator, events );
+
+        final String fname = "/path/to/my/file.txt";
+        Location loc = new SimpleLocation( "remote:foo/com" );
+        ConcreteResource resource = new ConcreteResource( loc, fname );
+        CacheProvider get = router.getRoutedProvider( resource );
```

Comment:
Okay, I see how this is used now. For testing, I'd prefer if you reduce scope to protected in the main implementation, then provide a TestRoutingCacheProvider class inside this test class...that can override the getRoutedProvider() method with public scope, and delegate to it, like this:

```
private static final class TestProvider extends RoutingCacheProviderWrapper{
    public CacheProvider getRoutedProvider( ConcreteResource resource ){
        return super.getRoutedProvider(resource);
    }
}
```

If this is internal and final, it can't be abused. Also, it allows the selection method to remain internal to the routing provider, to prevent a leaky abstraction.